cpp/src/arrow/util/align_util_test.cc

```diff
@@ -59,6 +63,18 @@ void CheckBitmapWordAlign(const uint8_t* data, int64_t bit_offset, int64_t lengt
   }
 }
 
+arrow::Result<std::shared_ptr<Array>> CreateUnalignedArray(std::shared_ptr<Array> array) {
+  BufferVector sliced_buffers(array->data()->buffers.size(), nullptr);
+  for (std::size_t i = 0; i < array->data()->buffers.size(); ++i) {
+    if (array->data()->buffers[i]) {
+      sliced_buffers[i] = array->data()->buffers[i]->CopySlice(0, 2).MoveValueUnsafe();
+    }
+  }
+  auto sliced_array_data =
+      ArrayData::Make(array->type(), array->length(), std::move(sliced_buffers));
```

Comment:
The length should be smaller here correct?  Since your arrays are all `uint8` you should decrease the length to match the slice.